File: static/viewer/scene/figurvideo_stuecke.js
/**
 * Figurvideostuecke — die Kleidungsstücke der Szene für den Server-Weg.
 *
 * WARUM (11.09.2026): Bis dahin kamen NUR GarmentCode-Stücke ins Server-
 * Video, über ihre `_sim_rig.json`. Vorlagen-Cloth (`tpl_`), Garderobe
 * (`gar_`) und MakeHuman-Proxys (`mh_`) haben keine solche Datei — die
 * Figur kam ohne sie an. Was in der Szene an der Figur hängt und gehäutet
 * ist, geht jetzt SO mit, wie es hier steht: Punkte, Dreiecke, Hautgewichte
 * — dieselben Daten, die der Shader benutzt.
 *
 * BINÄR, NICHT JSON: Ein T-Shirt hat 10.814 Punkte; als Zahlenliste in JSON
 * wären das rund 3 MB je Stück, als Puffer 640 KB. Aufbau je Stück, ohne
 * Kopf — die Zahlen stehen im Auftrag:
 *
 *     Float32[n*3]  Punkte (Lage der Figur, Y oben — der Server dreht)
 *     Uint32 [m*3]  Dreiecke
 *     Uint16 [n*4]  Knochennummern     (Reihenfolge: `knochen` im Auftrag)
 *     Float32[n*4]  Gewichte
 *
 * DIE KNOCHENNAMEN kommen aus `skinWeightData.bone_names` (mit Punkten,
 * `DEF-spine.004`), nicht aus `bone.name`: Three.js ersetzt dort Punkte
 * durch Unterstriche, und `DEF-upper_arm_L` lässt sich nicht eindeutig
 * zurückübersetzen. Der Aufrufer reicht sie herein — so braucht dieses
 * Modul weder `state.js` noch Three.js und läuft in Node
 * (`test_js_figurvideo_stuecke.py`).
 */
export class Figurvideostuecke {

    /** Alle gehäuteten Stücke der Figur als Pakete — und die Namen derer,
     *  die starr hängen und deshalb nicht mitkommen. */
    static sammeln(inst) {
        const skelett = inst?.bodyMesh?.skeleton;
        const stuecke = [], starre = [];
        for (const [schluessel, netz] of Object.entries(inst?.clothMeshes || {})) {
            const paket = (netz?.isSkinnedMesh && netz.skeleton === skelett)
                ? Figurvideostuecke.paket(netz, schluessel) : null;
            if (paket) stuecke.push(paket); else starre.push(schluessel);
        }
        return { stuecke, starre };
    }

    /** Ein Netz als Paket, oder null, wenn ihm Gewichte fehlen. */
    static paket(netz, schluessel) {
        const geo = netz.geometry;
        const pos = geo.getAttribute('position');
        const si = geo.getAttribute('skinIndex');
        const sw = geo.getAttribute('skinWeight');
        if (!pos || !si || !sw) return null;
        const n = pos.count;
        // Punkte in der Lage der FIGUR: die eigene Matrix des Netzes gehört
        // dazu (ein GarmentCode-Stück bindet mit `netz.matrix`).
        netz.updateMatrix();
        const m = netz.matrix.elements;
        const punkte = new Float32Array(n * 3);
        for (let i = 0; i < n; i++) {
            const x = pos.getX(i), y = pos.getY(i), z = pos.getZ(i);
            punkte[3 * i] = m[0] * x + m[4] * y + m[8] * z + m[12];
            punkte[3 * i + 1] = m[1] * x + m[5] * y + m[9] * z + m[13];
            punkte[3 * i + 2] = m[2] * x + m[6] * y + m[10] * z + m[14];
        }
        let dreiecke;
        if (geo.index) {
            dreiecke = Uint32Array.from(geo.index.array);
        } else {
            dreiecke = new Uint32Array(n);
            for (let i = 0; i < n; i++) dreiecke[i] = i;
        }
        const nummern = new Uint16Array(n * 4), gewichte = new Float32Array(n * 4);
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < 4; j++) {
                nummern[4 * i + j] = Math.round(si.getComponent(i, j));
                gewichte[4 * i + j] = sw.getComponent(i, j);
            }
        }
        const blob = new Blob([punkte.buffer, dreiecke.buffer, nummern.buffer, gewichte.buffer]);
        return {
            name: netz.userData?.beschriftung || netz.name || schluessel,
            schluessel, farbe: Figurvideostuecke.farbe(netz),
            punkte: n, dreiecke: dreiecke.length / 3, blob,
        };
    }

    /** Die Farbe des Materials als `#rrggbb` — beim Körper-Array das erste. */
    static farbe(netz) {
        const mat = Array.isArray(netz.material) ? netz.material[0] : netz.material;
        return mat?.color?.getHexString ? `#${mat.color.getHexString()}` : null;
    }

    /**
     * Das Formular für `POST /api/animation/video/`: der Auftrag als JSON
     * (mit der Stückliste und den Knochennamen), je Stück ein Paket.
     * `knochen` sind die Namen in Skelettreihenfolge (`bone_names`).
     */
    static formular(auftrag, stuecke, knochen) {
        const daten = new FormData();
        daten.append('auftrag', JSON.stringify({
            ...auftrag, knochen,
            stuecke: stuecke.map((s, i) => ({
                name: s.name, farbe: s.farbe, punkte: s.punkte,
                dreiecke: s.dreiecke, datei: `stueck_${i}`,
            })),
        }));
        stuecke.forEach((s, i) => daten.append(`stueck_${i}`, s.blob, `stueck_${i}.bin`));
        return daten;
    }
}

File: core/dienste/figurvideostuecke.py
# -*- coding: utf-8 -*-
"""Figurvideostuecke — Kleidungsstuecke aus der Szene fuer den Filmlauf.

WARUM (11.09.2026): Der Server-Weg nahm nur GarmentCode-Stuecke mit, ueber
ihre `_sim_rig.json`. Vorlagen-Cloth, Garderobe und MakeHuman-Proxys haben
keine solche Datei — die Figur kam im Video ohne sie an. Jetzt schickt der
Browser jedes gehaeutete Netz so, wie es in der Szene haengt
(`scene/figurvideo_stuecke.js`), und hier wird daraus je Stueck eine
`.npz`, die `ModelPhysik/figurnetze.py` wie eine Rig-Datei liest.

Aufbau eines Pakets (ohne Kopf, die Zahlen stehen im Auftrag):

    Float32[n*3]  Punkte, Lage der Figur, Y OBEN (Three.js)
    Uint32 [m*3]  Dreiecke
    Uint16 [n*4]  Knochennummern in der Reihenfolge von `knochen`
    Float32[n*4]  Gewichte

DIE ACHSEN WERDEN HIER GEDREHT: Die Szene rechnet Y oben, der Filmlauf in
Blender-Lage (Z oben) — `blender = (x, -z, y)`, die Umkehrung von
`blenderToThreeCoords`. Ein Stueck ohne diese Drehung laege im Video flach
am Boden neben der Figur (dieselbe Falle wie am 06.09.2026 beim Anziehen).
"""
import os

import numpy as np


class Figurvideostuecke:
    """Pakete aus dem Formular -> `.npz` je Stueck im Auftragsordner."""

    #: Standardfarbe, wenn das Material keine nennt (Blau wie bisher).
    FARBE = [0.29, 0.44, 0.72]

    @classmethod
    def ablegen(cls, dateien, liste, knochen, ordner):
        """Schreibt jedes Stueck und gibt `[{name, pfad, farbe}]` zurueck.

        `dateien` ist `request.FILES`, `liste` die Stueckliste des Auftrags
        (`name, farbe, punkte, dreiecke, datei`), `knochen` die Namensliste
        des Skeletts in der Reihenfolge der Knochennummern.
        """
        if not liste:
            return []
        if not knochen:
            raise ValueError('Stuecke ohne Knochenliste.')
        ziel = os.path.join(ordner, 'stuecke')
        os.makedirs(ziel, exist_ok=True)
        aus = []
        for nummer, meta in enumerate(liste):
            name = str(meta.get('name') or 'Stueck %d' % (nummer + 1))
            datei = dateien.get(str(meta.get('datei') or ''))
            if datei is None:
                raise ValueError('Stueck „%s" ohne Daten.' % name)
            pfad = os.path.join(ziel, '%02d.npz' % nummer)
            cls._schreiben(pfad, datei.read(), int(meta.get('punkte') or 0),
                           int(meta.get('dreiecke') or 0), knochen, name)
            aus.append({'name': name, 'pfad': pfad,
                        'farbe': cls.farbe(meta.get('farbe'))})
        return aus

    @classmethod
    def _schreiben(cls, pfad, roh, n, m, knochen, name):
        erwartet = n * 12 + m * 12 + n * 8 + n * 16
        if n <= 0 or m <= 0 or len(roh) != erwartet:
            raise ValueError('Stueck „%s": %d Bytes, erwartet %d '
                             '(%d Punkte, %d Dreiecke).'
                             % (name, len(roh), erwartet, n, m))
        punkte = np.frombuffer(roh, dtype=np.float32, count=n * 3
                               ).reshape(n, 3)
        versatz = n * 12
        dreiecke = np.frombuffer(roh, dtype=np.uint32, count=m * 3,
                                 offset=versatz).reshape(m, 3)
        versatz += m * 12
        nummern = np.frombuffer(roh, dtype=np.uint16, count=n * 4,
                                offset=versatz).reshape(n, 4)
        versatz += n * 8
        gewichte = np.frombuffer(roh, dtype=np.float32, count=n * 4,
                                 offset=versatz).reshape(n, 4)
        if dreiecke.size and int(dreiecke.max()) >= n:
            raise ValueError('Stueck „%s": Dreieck zeigt auf Punkt %d von %d.'
                             % (name, int(dreiecke.max()), n))
        belegt = nummern[gewichte > 0]
        if belegt.size and int(belegt.max()) >= len(knochen):
            raise ValueError('Stueck „%s": Knochennummer %d, aber nur %d '
                             'Namen.' % (name, int(belegt.max()), len(knochen)))
        # Three.js (Y oben) -> Blender (Z oben).
        blender = np.column_stack([punkte[:, 0], -punkte[:, 2], punkte[:, 1]])
        np.savez(pfad, punkte=blender.astype(np.float64),
                 dreiecke=dreiecke.astype(np.int64),
                 skin_index=nummern.astype(np.int32),
                 skin_weight=gewichte.astype(np.float64),
                 knochen=np.array([str(k) for k in knochen]))

    @classmethod
    def farbe(cls, wert):
        """`#rrggbb` -> [r, g, b] in 0..1; sonst die Standardfarbe."""
        try:
            farbe = str(wert or '').lstrip('#')
            if len(farbe) == 6:
                return [int(farbe[i:i + 2], 16) / 255.0 for i in (0, 2, 4)]
        # stumm gewollt: ein unlesbarer Farbwert aus dem Browser faellt auf die Standardfarbe
        except (TypeError, ValueError):
            pass
        return list(cls.FARBE)
